src/sage/geometry/cone_critical_angles.py

```diff
@@ -0,0 +1,998 @@
+      for the cone `P`
+
+    - ``hs`` -- a linearly independent list of unit-norm generators
+      for the cone `Q`
+
+    OUTPUT:
+
+    A tuple containing four elements, in order:
+
+    - The matrix `M` described in Proposition 6
+
+    - The minimal entry in the matrix `M`
+
+    - A vector in ``gs`` that achieves that minimal inner product
+      along with the next element of the tuple
+
+    - A vector in ``hs`` that achieves the minimal inner product
+      along with the previous element in the tuple
+
+    EXAMPLES::
+
+        sage: from sage.geometry.cone_critical_angles import compute_gevp_M
+        sage: P = Cone([ (1,2,0), (3,4,0) ])
+        sage: Q = Cone([ (-1,4,1), (5,-2,-1),  (-1,-1,5) ])
+        sage: gs = [g.change_ring(QQ) for g in P]
+        sage: hs = [h.change_ring(QQ) for h in Q]
+        sage: M = compute_gevp_M(gs, hs)[0]
+        sage: all( M[i][j] == gs[i].inner_product(hs[j])
+        ....:       for i in range(P.nrays())
+        ....:       for j in range(Q.nrays()) )
+        True
+
+    TESTS:
+
+    The products `(G_{I})^{T}H_{J}` correspond to
+    submatrices of the "generalized Gram matrix" `M` in Proposition
+    6. Note that SageMath does (row,column) indexing but [Or2020]_
+    does (column,row) indexing::
+
+        sage: from sage.geometry.cone_critical_angles import (
+        ....:   _random_admissible_cone,
+        ....:   compute_gevp_M,
+        ....:   gevp_licis)
+        sage: n = ZZ.random_element(1,4)
+        sage: n = ZZ.random_element(1,8) # long time
+        sage: P = _random_admissible_cone(ambient_dim=n)
+        sage: Q = _random_admissible_cone(ambient_dim=n)
+        sage: gs = [g.change_ring(QQ) for g in P]
+        sage: hs = [h.change_ring(QQ) for h in Q]
+        sage: M = compute_gevp_M(gs,hs)[0]
+        sage: f = lambda i,j: gs[i].inner_product(hs[j])
+        sage: expected_M = matrix(QQ, P.nrays(), Q.nrays(), f)
+        sage: M == expected_M
+        True
+        sage: G = matrix.column(gs)
+        sage: H = matrix.column(hs)
+        sage: def _test_indexing(I,J):
+        ....:      G_I = G[range(n), I]
+        ....:      H_J = H[range(n), J]
+        ....:      return (G_I.transpose()*H_J == M[I,J]
+        ....:              and
+        ....:              H_J.transpose()*G_I == M.transpose()[J,I])
+        sage: G_index_sets = list(gevp_licis(G))
+        sage: H_index_sets = list(gevp_licis(H))
+        sage: all( _test_indexing(I,J) for I in G_index_sets
+        ....:                          for J in H_index_sets )
+        True
+    """
+    min_u = gs[0]
+    min_v = hs[0]
+    min_ip = min_u.inner_product(min_v)
+
+    M = []
+    for i in range(len(gs)):
+        M_i = []
+        for j in range(len(hs)):
+            M_i.append(gs[i].inner_product(hs[j]))
+            if (M_i[j] < min_ip):
+                min_ip = M_i[j]
+                min_u = gs[i]
+                min_v = hs[j]
```

Comment:
```suggestion
        M_i = [None] * len(hs)
        for v in hs:
            val = M_i.append(gs[i].inner_product(v))
            if val < min_ip:
                min_ip = val
                min_u = gs[i]
                min_v = v
```
Similar such changes.

Reply:
list append doesn't return a value... or am I missing something? I could still assign `M_i[j]` to a variable to avoid re-indexing but surely by itself that's overkill.

Reply:
Ah, sorry, that was a mistake. It should have been
```python
for j, v in enumerate(hs):
    val = gs[i].inner_product(v)
    M_i[j] = val

Reply:
Since I'm keeping `M = []` and the append, I was able to omit the indices entirely. Looks like a small improvement, thanks!